packages/appium/docs/en/ecosystem/build-drivers.md

```diff
@@ -2,4 +2,617 @@
+Appium's `BaseDriver` (see below).
+
+### Extend Appium's `BaseDriver` class
+
+Ultimately, your driver is much easier to write because most of the hard work of implementing the
+WebDriver protocol and handling certain common logic is taken care of already by Appium. This is
+all encoded up as a class which Appium exports for you to use, called `BaseDriver`. It is exported
+from `appium/driver`, so you can use one of these styles to import it and create your *own* class
+that extends it:
+
+```js
+import {BaseDriver} from 'appium/driver';
+// or: const {BaseDriver} = require('appium/driver');
+
+export class MyDriver extends BaseDriver {
+}
+```
+
+### Make your driver available
+
+That's basically it! With a Node.js package exporting a driver class and with a correct Appium
+manifest, you've got yourself an Appium driver! Now it doesn't *do* anything, but you can load it
+up in Appium, start and stop sessions with it, etc...
+
+To make it available to users, you could publish it via NPM. When you do so, your driver will be
+isntallable via the Appium CLI:
+
+```
+appium driver install --source=npm <driver-package-on-npm>
+```
+
+### Developing your driver
+
+How you develop your driver is up to you. It is convenient, however, to run it from within Appium
+without having to do lots of publishing and installing. The most straightforward way to do this is
+to include the most recent version of Appium as a `devDependency`, and then also your own driver,
+like this:
+
+```json
+{
+    "devDependencies": {
+        ...,
+        "appium": "^2.0.0",
+        "your-driver": "file:.",
+        ...
+    }
+}
+```
+
+Now, you can run Appium locally (`npx appium`), and because your driver is listed as a dependency
+alongside it, it will be automatically "installed" and available. You can design your e2e tests
+this way, or if you're writing them in Node.js, you can simply import Appium's start server methods
+to handle starting and stopping the Appium server in Node.
+
+## Standard driver implementation ideas
+
+These are things you will probably find yourself wanting to do when creating a driver.
+
+### Set up state in a constructor
+
+If you define your own constructor, you'll need to call `super` to make sure all the standard state
+is set up correctly:
+
+```js
+constructor(opts) {
+    super(opts);
+    // now do your own thing
+}
+```
+
+### Define and validate accepted capabilities
+
+You can define your own capabilities and basic validation for them. Users will always be able to
+send in capabilities that you don't define, but if they send in capabilities you have explicitly
+defined, then Appium will validate that they are of the correct type (and will check for the
+presence of required capabilities).
+
+If you want to turn capability validation off entirely, set `this.shouldValidateCaps` to `false` in
+your constructor.
+
+To give Appium your validation constraints, set `this.desiredCapConstraints` to a validation object
```

Comment:
ugh, need schemas.

Reply:
yeah, want to make another schema system lol?

Reply:
luckily I can reuse a ton of it.  

should get done, but less of a high priority and more of a "fun" thing for me